PocketKit/Tests/SharedPocketKitTests/SubscriptionsStackTests.swift

```diff
@@ -0,0 +1,109 @@
+        subscriptionMessage = message
+    }
+}
+
+class SubscriptionsStackTests: XCTestCase {
+    var firstSet: Set<AnyCancellable>!
+    var secondSet: Set<AnyCancellable>!
+
+    func testPushSubscriptions() {
+        var stack = SubscriptionsStack()
+
+        let firstTester = SubscriptionTester()
+        firstTester.updateReceivedExpectation = expectation(description: "first update received")
+
+        firstSet = Set<AnyCancellable>()
+        firstTester.$subscriptionMessage
+            .dropFirst()
+            .sink { message in
+                XCTAssertEqual(message, "firstUpdate")
+            }
+            .store(in: &firstSet)
+
+        stack.push(firstSet)
+        firstSet = nil
+
+        let secondTester = SubscriptionTester()
+        secondTester.updateReceivedExpectation = expectation(description: "second update received")
+
+        secondSet = Set<AnyCancellable>()
+        secondTester.$subscriptionMessage
+            .dropFirst()
+            .sink { message in
+                XCTAssertEqual(message, "secondUpdate")
+            }
+            .store(in: &secondSet)
+        stack.push(secondSet)
+        secondSet = nil
+
+        firstTester.updateMessage("firstUpdate")
+        secondTester.updateMessage("secondUpdate")
+
+        wait(for: [firstTester.updateReceivedExpectation!, secondTester.updateReceivedExpectation!], timeout: 1)
+    }
+
+    func testPopSubscriptions() {
+        var stack = SubscriptionsStack()
+        var callCount = 0
+
+        let firstTester = SubscriptionTester()
+
+        firstSet = Set<AnyCancellable>()
+        firstTester.$subscriptionMessage
+            .dropFirst()
+            .sink { message in
+                callCount += 1
+            }
+            .store(in: &firstSet)
+
+        stack.push(firstSet)
+        firstSet = nil
+
+        let secondTester = SubscriptionTester()
+
+        secondSet = Set<AnyCancellable>()
+        secondTester.$subscriptionMessage
+            .dropFirst()
+            .sink { message in
+                callCount += 1
+            }
+            .store(in: &secondSet)
+        stack.push(secondSet)
+        secondSet = nil
+
+        firstTester.updateMessage("whateverMessage")
+        secondTester.updateMessage("whateverMessage")
+
+        // two updates should be received, callCount should be equal to 2
+        XCTAssertEqual(callCount, 2)
+        stack.pop()
+
+        firstTester.updateMessage("whateverMessage")
+        secondTester.updateMessage("whateverMessage")
+
+        // second set should be popped, thus, only 1 update should be received
+        // callCount should be equal to 3
+        XCTAssertEqual(callCount, 3)
+        stack.pop()
+
+        firstTester.updateMessage("whateverMessage")
+        secondTester.updateMessage("whateverMessage")
+
+        // no more updates should be received, callCount should be still equal to 3
+        XCTAssertEqual(callCount, 3)
+    }
```

Comment:
do we want to check for the tester expectations in this test similar to what we do in line 56?

Reply:
we could, but it would be redundant since we already check those in the previous test. Unless we have a way to validate that an expectation was not fulfilled, to enforce that the popped subscription did not trigger an update? I am not aware if we can do that

Reply:
what about using `expectation.isInverted = true` to validate that it wasn't fulfilled? but I'm okay with this, just a passing thought!